### ISSUE TITLE
vips8 python doesn't seem to work as documented

### ISSUE THREAD

**Issue by banks** [jcupitt/libvips#302]
Hi I'm trying to evaluate libvips as a potential for building a new engine for thumbor (https://github.com/thumbor/thumbor).

I realise that the new python binding is still a WIP and packaging is not complete yet, but I thought I would press on as the readme and comments on tickets and in libvips wiki docs seem to imply that it's perfectly usable once installed.

I finally managed to get it installed on a debian jessie based system (actually Docker container) partly aided by scraps of info in readme, docs, and tickets and partly trial and error.

So I can now import Vips from gi.repository just fine!

But then I started trying to actually program with it...

Even stuff I copied from the examples in the master repo don't seem to work. `Vips.Image.new_from_file()` function doesn't even exist, `im.width` doesn't exist etc. I got around those by inspecting the `Vips.Image` object and discovered that while that new method isn't there `new_from_file_RW()` is so for now I used that and `get_width()`...

Also, the Vips.py file that was installed when i compiled libvips *does* contain code for new_from_file:

```
# cat /usr/lib/python2.7/site-packages/gi/overrides/Vips.py | grep new_from_file
def vips_image_new_from_file(cls, vips_filename, **kwargs):
    logging.debug('Image.new_from_file: loader = %s' % loader)
setattr(Vips.Image, 'new_from_file', vips_image_new_from_file)
```

But if I try to call it from python I get 

```
    im = Vips.Image.new_from_file(input)
AttributeError: type object 'Image' has no attribute 'new_from_file'
```

And dumping the attributes of Image object reveals:

https://gist.github.com/banks/6ee36c19244ce50f7402

Anyway I'll use methods I can see in dir only for now...

---

But then I want to actually do something with my image (resize it) and reading docs which describe how the C API operation methods like `resize` map to `image.resize()`. But Image.resize is not defined either... I see some examples that use `Vips.Operation` objects directly so perhaps that is what I need...

But for example this code: https://github.com/jcupitt/libvips/blob/master/python/example/try10.py

Fails for me since _none_ of those methods exist in the current version of the binding I just built...

I realise this is a slightly unhelpful ticket given that this binding is not complete but can you help me out - Am I doing something wrong or is http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/using-from-python.html and the examples in the repo out of date?

If you have any other resource I can look at that would help me use the new API without reverse engineering it all from source/introspection then please let me know - would really love to try this great project.

Thanks for all your effort on this.

**Comment by jcupitt** [comment]
Hi Paul, sorry you're having trouble.

The binding is complete and does all work, it's just annoying to install from source. It sounds like your overrides (in Vips.py) are not being picked up. 

Here's what happens when Python executes `from gi.repository import Vips`:

1. It searches for a typelib called `Vips-x.y.typelib`. This is a binary file generated during libvips build by introspection of the `libvips.so` library plus scanning of the C headers. It lists all the API entry points, all the types the library uses, and has an extra set of hints for object ownership and reference counting. The typelib is searched for in `/usr/lib/gi-repository-1.0` and along the path in the environment variable `GI_TYPELIB_PATH`. 

2. It uses the typelib to make a basic binding for libvips. It's just the C API with a little very light mangling, so for example the enum member `VIPS_FORMAT_UCHAR` of the enum `VipsBandFormat` becomes `Vips.BandFormat.UCHAR`. It sounds like this much is working for you.

3. The binding you get is rather unfriendly, so it also loads a set of overrides from `Vips.py` in `/usr/lib/python2.7/dist-packages/gi/overrides` (on my system at least). If you're using python3, it's `/usr/lib/python3/dist-packages/gi/overrides`. Unfortunately, as far as I know, there is no way to extend this search using environment variables. You MUST copy `Vips.py` (and `Vips.pyc` I guess) to exactly this directory. If you install vips via a package manager this will happen automatically, since vips and pygobject will have been built to the same prefix, but if you are installing vips from source and the prefix does not match, it will not be installed for you. I've mailed pygobject and asked about this issue, but no response so far.

4. Finally, `Vips.py` makes the rest of the binding. In fact, `Vips.py` (as you've seen) makes almost all the binding: it defines `__getattr__` on `Vips.Image` and binds at runtime by searching libvips for an operation of that name. It defines all the docstrings too, so you should be able to use `help()` to find stuff.

There's a fairly large test suite in `libvips/test` done in vips8 Python which passes and which does not leak, so I promise it does all work :-)

**Comment by banks** [comment]
@jcupitt Thanks for taking time for such a detailed response - I was indeed not clear enough that Vips.py overrides were not essential for import to work so I went through enough attempts to get deps sorted and things in right place that I assumed by time it imported without error it must all be installed correctly.

FWIW I believe it was the compile form source (i.e. `make install`) that decided to copy Vips.py to site-packages and not dist-packages. I'm not sure why but I assumed it had done the right thing. In my environment (debian) it seems dist-packages is only thing in sys.path by default (not site-packages) if you are using a distribution python rather than a from-source one.

This worked great once I stopped being dumb and copied it into dist-packages as the comments say.
